### PR TITLE
26 containerize ingestion

### LINKED ISSUE #26: Containerize ingestion service

### PR DESCRIPTION
**Pull request by ghostaccount737**
Closes #26 and #25

### CODE REVIEW

**Comment by PVinay737** [comment]
Need to change configs

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM python:3.5
+
+MAINTAINER Vinay
+RUN apt-get update -y && \
+   apt-get install -y python3-dev python3-pip
```
Comment (lastlegion) [inline Dockerfile]: isnt' this part of python base image?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+WORKDIR /inference/app
+COPY . /inference/app
```
Comment (lastlegion) [inline Dockerfile]: copy requirements.txt seperately

**Comment by lastlegion** [comment]
Hi @PVinay737 please check this article: https://pythonspeed.com/articles/dockerizing-python-is-hard/. It has a lot of useful information on dockerizing python applications and covers common mistakes that a lot of developers make.

File: inference/run.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
 from app import app
 
 if __name__ == "__main__":
-    app.run()
+    app.run(host='0.0.0.0', debug=True)
```
Comment (lastlegion) [inline inference/run.py]: make configurable.

**Comment by PVinay737** [comment]
@lastlegion will work on updating Dockerization to the best practices and add it here by tomorrow. Whenever you are free kindly groom the next issues that I can work upon once this PR is merged.

File: Dockerfile (hunk 1 above)
Comment (PVinay737) [reply Dockerfile]: kindly check the update

**Comment by PVinay737** [comment]
Let me know the next issues I can look at once this is closed

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.7.3-stretch
+
+MAINTAINER Vinay
+RUN apt-get update -y
+WORKDIR /inference/app
+COPY requirements.txt /inference/app
+RUN pip install -r /inference/app/requirements.txt
+EXPOSE 5000
```
Comment (lastlegion) [inline Dockerfile:10]: where is the code being copied?

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+web:
+    build: .
+    command: python -u inference/run.py
+    ports:
+        - "5000:5000"
+    volumes:
+        - .:/inference/app
+    links:
+        - vizrec
+vizrec:
```
Comment (lastlegion) [inline docker-compose.yml]: call it db?

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    image: mongo:latest
```
Comment (lastlegion) [inline docker-compose.yml]: pin version

File: inference/config.py [hunk 7]
```diff
@@ -28,3 +29,8 @@ class TestingConfig(Config):
     UPLOAD_FOLDER = os.path.join(APP_ROOT, 'app', 'static')
     ALLOWED_EXTENSIONS = ["JSON"]
     MAX_CONTENT_LENGTH = 16 * 1024 * 100024
+    # URI = 'mongodb://127.0.0.1:5000'
+    CLIENT = pymongo.MongoClient('172.17.0.1', 27017)
```
Comment (lastlegion) [inline inference/config.py]: why is default `172.17.0.1`, consider using service name from docker compose file?

File: inference/app/__init__.py [hunk 8]
```diff
@@ -36,19 +37,19 @@ def allowed_filesize(filesize):
...
+        if "filesize" in request.cookies:
+            if not allowed_filesize(request.cookies["filesize"]):
+                return handle_bad_request('e')
+        f1 = request.files["file"]
+        if allowed_file(f1.filename):
+            full_filename = os.path.join(
+                app.config['UPLOAD_FOLDER'],
+                secure_filename(
+                    f1.filename))
+            f1.save(full_filename)
```
Comment (lastlegion) [inline inference/app/__init__.py]: check for exceptions

File: inference/tests/test_app.py [hunk 9]
```diff
@@ -1,8 +1,22 @@
...
+    assert response.status_code == 400
+    response1 = client.get("/upload_file", data=json.dumps(dict(f='f')))
+    assert response1.status_code == 405
+    valids = ["json"]
+    data = {'file': '1.json'}
+    field = data['file']
+    assert(glob.glob(os.getcwd() + '/**/' + field, recursive=True))
+    assert((data['file'].split('.'))[1] in valids)
+    try:
+        client.post('/upload_file', data=data)
```
Comment (lastlegion) [inline inference/tests/test_app.py]: what are we asserting after this post?

File: inference/app/__init__.py (hunk 8 above)
Comment (lastlegion) [inline inference/app/__init__.py]: can this be loaded to mongo directly without writing to disk?

**Comment by lastlegion** [comment]
Please rename the directory to `ingestion`

File: inference/app/__init__.py (hunk 8 above)
Comment (ghostaccount737) [reply inference/app/__init__.py]: Would  `handle_bad_request`  be fine ?

File: inference/tests/test_app.py (hunk 9 above)
Comment (PVinay737) [reply inference/tests/test_app.py]: checking for valid json files wheras the above one is not for .json files

Comment (PVinay737) [reply inference/tests/test_app.py]: checking for valid JSON and not .json files

Comment (ghostaccount737) [reply inference/tests/test_app.py]: Asserting for.JSON file. Is that the intent?

File: inference/app/__init__.py (hunk 8 above)
Comment (ghostaccount737) [reply inference/app/__init__.py]: something like mongofiles ?

Comment (lastlegion) [reply inference/app/__init__.py]: no just inserting the JSON into mongo directly using something like `insert_many` http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.insert_many?

Comment (PVinay737) [reply inference/app/__init__.py]: Isn't this the case of multiple files? Can we discuss this in a newer issue as this one already has two of them

Comment (ghostaccount737) [reply inference/app/__init__.py]: Maybe newer issue ?

**Review by lastlegion — APPROVED** [review]
Still scope for lots of improvement, but approving to speed up the process.

**Pull request merged**